### PR TITLE
Rec: Move check implementation to NodeCheck

### PR DESCRIPTION
**Pull request by KnapSac**
To reduce code duplication, the `NodeCheck` class is introduced, which implements the `Check` method for checks which can have sub-checks.

Also implemented `NotCheck`.

### CODE REVIEW

File: PatternPal/PatternPal.Core/Checks/MethodCheck.cs [hunk 1]
```diff
@@ -1,100 +1,29 @@
...
     {
-        MatchedEntities = new List<INode>();
-        _checks = checks;
     }
 
-    internal Func<List<INode>> Result => () => MatchedEntities;
-
     /// <inheritdoc />
-    public override ICheckResult Check(
+    protected override IEntity GetType4TypeCheck(
```
Comment (ViciousDoormat) [inline PatternPal/PatternPal.Core/Checks/MethodCheck.cs:19]: Add a comment explaining that the type checked in the case of a method is the return type

File: PatternPal/PatternPal.Core/Checks/NodeCheck.cs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                    foreach (IMethod method in entity.GetMethods())
+                    {
+                        subCheckResults.Add(
+                            methodCheck.Check(
+                                ctx,
+                                method));
+                    }
+                    break;
+                }
+                case FieldCheck fieldCheck:
```
Comment (ViciousDoormat) [inline PatternPal/PatternPal.Core/Checks/NodeCheck.cs:94]: You have specified this for specifically a class. I would instead add a method GetFields which ClassCheck implements

Comment (ViciousDoormat) [reply PatternPal/PatternPal.Core/Checks/NodeCheck.cs:94]: same for constructorcheck and propertycheck,

**Review by ViciousDoormat — CHANGES_REQUESTED** [review]

File: PatternPal/PatternPal.Core/Checks/NodeCheck.cs (hunk 2 above)
Comment (KnapSac) [reply PatternPal/PatternPal.Core/Checks/NodeCheck.cs:94]: I don't think the added complexity is worth it in this case.

**Comment by KnapSac** [comment]
@ViciousDoormat did you rereview this PR? It seems there might have been an issue with automerge.

File: PatternPal/PatternPal.Core/Checks/NodeCheck.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    /// <returns>The <see cref="ICheckResult"/> of the <paramref name="nodeCheck"/>.</returns>
+    private static ICheckResult RunCheckWithMultipleMatches< T >(
+        RecognizerContext ctx,
+        IEnumerable< T > nodes,
+        NodeCheck< T > nodeCheck)
+        where T : INode
+    {
+        // Run the check on the nodes.
+        IList< ICheckResult > results = new List< ICheckResult >();
+        foreach (T method in nodes)
```
Comment (ViciousDoormat) [inline PatternPal/PatternPal.Core/Checks/NodeCheck.cs:172]: This should not be called "mehod"

File: PatternPal/PatternPal.Tests/Checks/NotCheckTests.cs [hunk 4]
```diff
@@ -9,82 +9,60 @@ namespace PatternPal.Tests.Checks;
...
-        NodeCheckResult expectedResult1 = new NodeCheckResult { Priority = Priority.Low, FeedbackMessage = "", ChildrenCheckResults = new List<ICheckResult> { leafModCheckResult1 } };
-
-        // get result of NotCheck
-        NodeCheckResult result1 = (NodeCheckResult)notCheck1.Check(ctx, classEntity);
-
-        Assert.AreEqual(expectedResult1.GetType(), result1.GetType());
-        Assert.AreEqual(expectedResult1.Priority, result1.Priority);
-        Assert.AreEqual(expectedResult1.ChildrenCheckResults.Count, result1.ChildrenCheckResults.Count);
-
+        CheckCollection checkCollection = Any(
```
Comment (ViciousDoormat) [inline PatternPal/PatternPal.Tests/Checks/NotCheckTests.cs:35]: Wouldn't it be better to put this in a ClassCheck instead of a CheckCollection

**Pull request merged**